File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string S, int k) {
        if(S.length() <= k)     return "0";
        
        stack<char>st;
        st.push(S[0]);
        
        for(int i = 1; i < S.length(); i++){
            while(k > 0 && !st.empty() && st.top() > S[i]){
                --k;
                st.pop();
            }
            st.push(S[i]);
        }
        
        while(k > 0 && !st.empty()){
            st.pop();
            --k;
        }
        
        S.clear();
        
        while(!st.empty()){
            S.push_back(st.top());
            st.pop();
        }
        
        while(S.length() > 0 && S.back() == '0')
            S.pop_back();
            
        reverse(S.begin(), S.end());
        
        return (S == ""  ? "0" : S);
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        int n = num.length();
        
        if(n == k)  return "0";
        
        Deque<Character> st = new ArrayDeque<>();
        st.push(num.charAt(0));
        
        int i = 1;
        
        while(i < n){
            while(!st.isEmpty() && k != 0 && st.peek() > num.charAt(i)){
                st.pop();
                --k;
            }
            
            st.push(num.charAt(i++));
        }
        
        StringBuilder res = new StringBuilder();
        
        while(!st.isEmpty()){
            res.append(st.peek());
            st.pop();
        }
        
        while(res.length() != 0 && res.charAt(res.length() - 1) == '0'){
            res.setLength(res.length() - 1);
        }
        
        res.reverse();
        
        if(res.length() == 0 || k >= res.length())   return "0";
        
        if(k != 0)  res.setLength(res.length() - k);
        
        return res.toString();
    }
}